### PR TITLE
#1245 Investigate Antlr grammar processing issues for the AQL

### LINKED ISSUE #1245: Investigate Antlr grammar processing issues for the AQL
# Task Description

We are getting the following warnings and need to look into the issue and resolve it:
```
[INFO] 
[INFO] --- antlr4-maven-plugin:4.7.1:antlr4 (aql-antlr-java) @ strongbox-aql ---
[INFO] ANTLR 4: Processing source directory /java/opensource/carlspring/strongbox/strongbox-aql/target/antlr4
[INFO] Processing grammar: org/carlspring/strongbox/aql/grammar/AQL.g4
[WARNING] warning(184): org/carlspring/strongbox/aql/grammar/AQL.g4:192:0: One of the token LAYOUT_COORDINATE_NPM_EXTENSION values unreachable. extension is always overlapped by token LAYOUT_COORDINATE_MAVEN_EXTENSION
[WARNING] /java/opensource/carlspring/strongbox/org/carlspring/strongbox/aql/grammar/AQL.g4 [192:0]: One of the token LAYOUT_COORDINATE_NPM_EXTENSION values unreachable. extension is always overlapped by token LAYOUT_COORDINATE_MAVEN_EXTENSION
[WARNING] warning(180): org/carlspring/strongbox/aql/grammar/AQL.g4:357:4: chars '\\' used multiple times in set [\\"\\']
[WARNING] /java/opensource/carlspring/strongbox/org/carlspring/strongbox/aql/grammar/AQL.g4 [357:4]: chars '\\' used multiple times in set [\\"\\']
[WARNING] warning(180): org/carlspring/strongbox/aql/grammar/AQL.g4:357:14: chars '\\' used multiple times in set [\\"\\'\r\n]
[WARNING] /java/opensource/carlspring/strongbox/org/carlspring/strongbox/aql/grammar/AQL.g4 [357:14]: chars '\\' used multiple times in set [\\"\\'\r\n]
[WARNING] warning(180): org/carlspring/strongbox/aql/grammar/AQL.g4:357:28: chars '\\' used multiple times in set [\\"\\']
[WARNING] /java/opensource/carlspring/strongbox/org/carlspring/strongbox/aql/grammar/AQL.g4 [357:28]: chars '\\' used multiple times in set [\\"\\']
```

# Acceptance Test

* [ ] Building the code with `mvn clean install -Dintegration.tests` still works.
* [ ] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [ ] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [ ] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @fuss86

### PR DESCRIPTION
**Pull request by timur27**
Fixes #1245.

### CODE REVIEW

**Comment by carlspring** [comment]
@timur27 ,

You don't need to make separate pulls for the same issue.

**Comment by timur27** [comment]
@sbespalov Created new test cases as per your advice. Kindly check. 
Thanks for your time!

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 1]
```diff
@@ -71,6 +71,99 @@ public void testLayoutSpecificKeywords()
...
+
+        AqlQueryParser aqlParser = new AqlQueryParser(query);
+        Map<Pair<Integer, Integer>, String> errorMap = null;
+        try
+        {
+            aqlParser.parseQuery();
+        }
+        catch (QueryParserException e)
+        {
+            errorMap = aqlParser.getErrors();
```
Comment (sbespalov) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: we don't expect exceptions here, so no need to catch anything

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 2]
```diff
@@ -71,6 +71,99 @@ public void testLayoutSpecificKeywords()
...
```
Comment (sbespalov) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: same as above

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 3]
```diff
@@ -71,6 +71,99 @@ public void testLayoutSpecificKeywords()
...
```
Comment (sbespalov) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: same here

**Comment by timur27** [comment]
Refreshed the PR. Kindly review

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 4]
```diff
@@ -71,6 +71,87 @@ public void testLayoutSpecificKeywords()
...
+        AqlQueryParser aqlParser = new AqlQueryParser(query);
+        Map<Pair<Integer, Integer>, String> errorMap = null;
+
+        aqlParser.parseQuery();
+
+
+        logger.info(String.format("Query [%s] parse tree:\n[%s]", query, aqlParser));
+
+        assertFalse(aqlParser.hasErrors());
+        assertNull(errorMap);
```
Comment (sbespalov) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: `errorMap` and this assert don't need any more as well

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 5]
```diff
@@ -71,6 +71,87 @@ public void testLayoutSpecificKeywords()
...
+
+        AqlQueryParser aqlParser = new AqlQueryParser(query);
+        Map<Pair<Integer, Integer>, String> errorMap = null;
+
+        aqlParser.parseQuery();
+
+        logger.info(String.format("Query [%s] parse tree:\n[%s]", query, aqlParser));
+
+        assertFalse(aqlParser.hasErrors());
+        assertNull(errorMap);
```
Comment (sbespalov) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: same as above

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 6]
```diff
@@ -71,6 +71,87 @@ public void testLayoutSpecificKeywords()
...
+        AqlQueryParser aqlParser = new AqlQueryParser(query);
+        Map<Pair<Integer, Integer>, String> errorMap = null;
+
+        aqlParser.parseQuery();
+
+
+        logger.info(String.format("Query [%s] parse tree:\n[%s]", query, aqlParser));
+
+        assertFalse(aqlParser.hasErrors());
+        assertNull(errorMap);
```
Comment (sbespalov) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: same as above

**Comment by timur27** [comment]
Thanks for comments. Refreshed the PR.

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 7]
```diff
@@ -71,6 +71,84 @@ public void testLayoutSpecificKeywords()
...
+    }
+
+
+    @Test
+    public void testValidSingleQuotedQuerySpecificChars()
+    {
+        String query = "repository: 'releases#1'";
+
+        AqlQueryParser aqlParser = new AqlQueryParser(query);
+        Map<Pair<Integer, Integer>, String> errorMap = null;
```
Comment (sbespalov) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: @timur27 could you please completely remove this `errorMap` variable as requested before
thanks :)

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 8]
```diff
@@ -71,6 +71,84 @@ public void testLayoutSpecificKeywords()
...
+        assertEquals(Pair.with(1, 21), errorPositionList.get(1));
+    }
+
+    @Test
+    public void testValidUnquotedStringQuery()
+    {
+        String query = "repository: releases_1";
+
+        AqlQueryParser aqlParser = new AqlQueryParser(query);
+        Map<Pair<Integer, Integer>, String> errorMap = null;
```
Comment (sbespalov) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: and here

File: strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java [hunk 9]
```diff
@@ -71,6 +71,84 @@ public void testLayoutSpecificKeywords()
...
+
+        assertFalse(aqlParser.hasErrors());
+
+    }
+
+
+    @Test
+    public void testValidSingleQuotedQuerySpecificChars()
+    {
+        String query = "repository: 'releases#1'";
```
Comment (carlspring) [inline strongbox-aql/src/test/java/org/carlspring/strongbox/aql/AqlParserTest.java]: Repositories cannot have such names with special characters. Would you mind changing this to `releases` (here and everywhere else across the pull request)? Thanks! :)

**Comment by sbespalov** [comment]
@carlpring, the thing is that we need to check the special characters
within quoted string here. If you don't like this name then we should think
about another with special character.

пн, 27 мая 2019 г., 18:24 Martin Todorov <notifications@github.com>:

**Comment by carlspring** [comment]
Oh, I see! Thanks for clearing that out! I've retracted my comment! :)

**Comment by timur27** [comment]
Ready to check. Thanks

**Comment by carlspring** [comment]
@sbespalov ,

Would you mind having another look?

**Pull request merged**